### PR TITLE
SALTO-1849: support app installation deploy

### PR DESCRIPTION
**Pull request by noamhammer**
_support app installation deploy_

---

_Additional context for reviewer_

---
_Release Notes_: 
_None_

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/zendesk-support-adapter/src/config.ts [hunk 1]
```diff
@@ -1343,7 +1366,7 @@ export const DEFAULT_TYPES: Record<string, configUtils.TypeDuckTypeConfig> = {
   // eslint-disable-next-line camelcase
   app_installations: {
     request: {
-      url: '/apps/installations',
+      url: '/apps/installations?include=app',
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/config.ts:1369]: nit:
```suggestion
      url: '/apps/installations',
      queryParams: {
        include: 'app',
      },
```

File: packages/zendesk-support-adapter/src/config.ts [hunk 2]
```diff
@@ -878,10 +878,33 @@ export const DEFAULT_TYPES: Record<string, configUtils.TypeDuckTypeConfig> = {
...
       sourceTypeName: 'app_installations__installations',
-      fieldsToHide: FIELDS_TO_HIDE.concat({ fieldName: 'id', fieldType: 'number' }),
+      fieldsToHide: FIELDS_TO_HIDE.concat(
+        { fieldName: 'id', fieldType: 'number' },
+        { fieldName: 'app' },
+      ),
       fieldsToOmit: FIELDS_TO_OMIT.concat({ fieldName: 'updated', fieldType: 'string' }),
-      idFields: ['settings.name'],
-      fileNameFields: ['settings.name'],
+      idFields: ['settings.name', 'app_id'],
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/config.ts:886]: The app_id is env friendly right?

File: packages/zendesk-support-adapter/src/filters/app.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+const INTERVAL_TIME = 5000
+
+const checkIfJobIsDone = async (
+  client: ZendeskClient, jobId: string, fullName: string
+): Promise<boolean> => {
+  const res = await client.getSinglePage({ url: `/apps/job_statuses/${jobId}` })
+  if (_.isArray(res.data)) {
+    throw new Error(`Got an invalid response for job status. Element: ${fullName}. Job ID: ${jobId}`)
+  }
+  if (res.data.status === 'failed') {
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/filters/app.ts:40]: Can you check here also for 'killed'? (or maybe there is a different way to check if the job was done and then you can check if it was done successfully?)

File: packages/zendesk-support-adapter/src/filters/app.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import { deployChange, deployChanges } from '../deployment'
+import ZendeskClient from '../client/client'
+
+export const APP_INSTALLATION_TYPE_NAME = 'app_installation'
+
+const { withRetry } = retry
+const { intervals } = retry.retryStrategies
+
+const MAX_RETRIES = 120
+const INTERVAL_TIME = 5000
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/filters/app.ts:31]: This is a pretty long timeout, can we make it shorter?

File: packages/zendesk-support-adapter/src/filters/app.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      strategy: intervals({
+        maxRetries: MAX_RETRIES,
+        interval: INTERVAL_TIME,
+      }),
+    }
+  )
+}
+
+/**
+ * Deploys business hours schedules and their intervals
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/filters/app.ts:60]: nit: update doc

File: packages/zendesk-support-adapter/src/filters/app.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      changes,
+      change => isAdditionOrModificationChange(change)
+        && (getChangeData(change).elemID.typeName === APP_INSTALLATION_TYPE_NAME),
+    )
+    const deployResult = await deployChanges(
+      relevantChanges,
+      async change => {
+        const fullName = getChangeData(change).elemID.getFullName()
+        const response = await deployChange(change, client, config.apiDefinitions, ['app', 'settings.title', 'settings_objects'])
+        if (isAdditionChange(change)) {
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/filters/app.ts:74]: updating and removing does not require waiting for the job to finish?

File: packages/zendesk-support-adapter/src/filters/app.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+const { withRetry } = retry
+const { intervals } = retry.retryStrategies
+
+const MAX_RETRIES = 120
+const INTERVAL_TIME = 5000
+
+const checkIfJobIsDone = async (
+  client: ZendeskClient, jobId: string, fullName: string
+): Promise<boolean> => {
+  const res = await client.getSinglePage({ url: `/apps/job_statuses/${jobId}` })
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/filters/app.ts:36]: nit: is there a value in adding a log here so we will see how much time it takes to install the app in customer accounts?

File: packages/zendesk-support-adapter/src/config.ts (hunk 2 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/config.ts:886]: Yes

File: packages/zendesk-support-adapter/src/filters/app.ts (hunk 6 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/filters/app.ts:74]: That's right - only the installation (which is the addition)

File: packages/zendesk-support-adapter/src/filters/app.ts (hunk 7 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/filters/app.ts:36]: I don't think so, we do get the exact time the installation was completed when in the result message (if it was successful), but we don't get the server start time. So I am not sure what is the value here.

File: packages/zendesk-support-adapter/src/filters/app.ts (hunk 4 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/filters/app.ts:31]: ok, I will make it 60 retries with 2 seconds for the interval. Do you think its fine?

File: packages/zendesk-support-adapter/src/config.ts [hunk 8]
```diff
@@ -1366,7 +1366,7 @@ export const DEFAULT_TYPES: Record<string, configUtils.TypeDuckTypeConfig> = {
   // eslint-disable-next-line camelcase
   app_installations: {
     request: {
-      url: '/apps/installations?include=app',
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/config.ts]: We don't need the `include=app`?

**Review by alonstern — APPROVED** [review]

File: packages/zendesk-support-adapter/src/config.ts (hunk 8 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/config.ts]: We needed it in the past but not anymore

**Pull request merged**